lib/Dialects/VectorExt/Transform/VectorWarpUtils.cpp

```diff
@@ -173,6 +176,119 @@ struct WarpOpElementwise : public OpRewritePattern<WarpSingleLaneOp> {
   }
 };
 
+/// A pattern that extracts vector.reduction ops from a WarpSingleLaneOp.
+/// The vector is reduced in parallel. Currently limited to vector<32x...>
+/// values. Every lane reduces two scalars, 5 times in a row.
+/// E.g.:
+/// ```
+/// %r = vector_ext.warp_execute_on_lane_0(%laneid) -> (f32) {
+///   %0 = "some_def"() : () -> (vector<32xf32>)
+///   %1 = vector.reduction "add", %0 : vector<32xf32> into f32
+///   vector_ext.yield %1 : f32
+/// }
+/// ```
+/// is lowered to:
+/// ```
+/// %0 = vector_ext.warp_execute_on_lane_0(%laneid) -> (vector<1xf32>) {
+///   %1 = "some_def"() : () -> (vector<32xf32>)
+///   vector_ext.yield %1 : vector<32xf32>
+/// }
+/// %a = vector.extract %0[0] : vector<1xf32>
+/// %r0, %s0 = gpu.shuffle down %e, %c16, %c32 : f32
+/// %a0 = arith.addf %a, %r0 : f32
+/// %r1, %s1 = gpu.shuffle down %e, %c8, %c32 : f32
+/// %a1 = arith.addf %a0, %r1 : f32
+/// ...
+/// %r4, %s4 = gpu.shuffle down %e, %c1, %c32 : f32
+/// %a4 = arith.addf %a3, %r4 : f32
+/// %r = gpu.shuffle idx %a4, %c0, %c32 : f32
+/// ```
+struct WarpOpReduction : public OpRewritePattern<WarpSingleLaneOp> {
+  using OpRewritePattern<WarpSingleLaneOp>::OpRewritePattern;
+
+  // Reduce the two given scalars.
+  static Value reduceScalars(OpBuilder &b, Location loc, Value v1, Value v2,
```

Comment:
This has a bad smell to it .. can we update core to use CombiningKind and reuse / expose a builder with CombiningKind ?